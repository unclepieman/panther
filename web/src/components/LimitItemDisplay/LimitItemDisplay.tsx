/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { Flex, Tooltip, Box } from 'pouncejs';
import NotificationJewel from 'Components/NotificationJewel';

interface LimitItemDisplayProps {
  /**
   * How many items should we show before we limit them
   */
  limit: number;

  /**
   * @ignore
   */
  children: React.ReactNode | React.ReactNode[];
}

const LimitItemDisplay: React.FC<LimitItemDisplayProps> = ({ limit, children }) => {
  const childrenCount = React.Children.count(children);
  if (childrenCount <= limit) {
    return children as any;
  }

  const childrenList = React.Children.toArray(children);
  const displayedChildren = childrenList.slice(0, limit);
  const hiddenChildren = childrenList.slice(limit);

  return (
    <React.Fragment>
      {displayedChildren}
      <Tooltip
        content={
          <Flex direction="column" spacing={1}>
            {hiddenChildren}
          </Flex>
        }
      >
        <Box>
          <NotificationJewel badge={`+${childrenCount - limit}`} />
        </Box>
      </Tooltip>
    </React.Fragment>
  );
};

export default React.memo(LimitItemDisplay);
