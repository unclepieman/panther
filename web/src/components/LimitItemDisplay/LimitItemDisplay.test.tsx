/**
 * Panther is a Cloud-Native SIEM for the Modern Security Team.
 * Copyright (C) 2020 Panther Labs Inc
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import { render, fireEvent } from 'test-utils';
import LimitItemDisplay from './LimitItemDisplay';

describe('LimitItemDisplay', () => {
  it('shows all the items if they are equal to or less than the limit', () => {
    const { getByText } = render(
      <LimitItemDisplay limit={2}>
        <div>One</div>
        <div>Two</div>
      </LimitItemDisplay>
    );

    expect(getByText('One')).toBeInTheDocument();
    expect(getByText('Two')).toBeInTheDocument();
  });

  it('shows less items if they are more than the limit and displays the rest in a tooltip', async () => {
    const { queryByText, findByText } = render(
      <LimitItemDisplay limit={2}>
        <div>One</div>
        <div>Two</div>
        <div>Three</div>
        <div>Four</div>
      </LimitItemDisplay>
    );

    expect(queryByText('Three')).not.toBeInTheDocument();
    expect(queryByText('Four')).not.toBeInTheDocument();

    fireEvent.mouseEnter(queryByText('+2'));

    expect(await findByText('Three')).toBeInTheDocument();
    expect(queryByText('Four')).toBeInTheDocument();
  });

  it('matches snapshot', async () => {
    const { container, getByText, findByText } = render(
      <LimitItemDisplay limit={2}>
        <div>One</div>
        <div>Two</div>
        <div>Three</div>
        <div>Four</div>
      </LimitItemDisplay>
    );
    expect(container).toMatchSnapshot();

    fireEvent.mouseEnter(getByText('+2'));
    expect(await findByText('Three')).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
