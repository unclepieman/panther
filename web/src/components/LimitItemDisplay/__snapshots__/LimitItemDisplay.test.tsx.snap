// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LimitItemDisplay matches snapshot 1`] = `


.panther-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 18px;
  height: 18px;
  background-color: #4f627c;
  border-radius: 99999px;
  font-size: 0.563rem;
  font-weight: 500;
  cursor: default;
}

<div>
  <header
    id="main-header"
  />
  <div>
    One
  </div>
  <div>
    Two
  </div>
  <div
    class="panther-1"
    data-reach-tooltip-trigger=""
  >
    <div
      class="panther-0"
    >
      +2
    </div>
  </div>
  <footer
    id="footer"
  />
</div>
`;

exports[`LimitItemDisplay matches snapshot 2`] = `


.panther-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 18px;
  height: 18px;
  background-color: #4f627c;
  border-radius: 99999px;
  font-size: 0.563rem;
  font-weight: 500;
  cursor: default;
}

<div>
  <header
    id="main-header"
  />
  <div>
    One
  </div>
  <div>
    Two
  </div>
  <div
    aria-describedby="tooltip--7"
    class="panther-1"
    data-reach-tooltip-trigger=""
  >
    <div
      class="panther-0"
    >
      +2
    </div>
  </div>
  <footer
    id="footer"
  />
</div>
`;
